src/folding/nova/circuits.rs

```diff
@@ -0,0 +1,275 @@
+where
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    _c: PhantomData<C>,
+    cmE: GC,
+    cmW: GC,
+}
+
+impl<C, GC> AllocVar<CommittedInstance<C>, CF2<C>> for CommittedInstanceE2Var<C, GC>
+where
+    C: CurveGroup,
+    GC: CurveVar<C, CF2<C>>,
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    fn new_variable<T: Borrow<CommittedInstance<C>>>(
+        cs: impl Into<Namespace<CF2<C>>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        f().and_then(|val| {
+            let cs = cs.into();
+
+            let cmE = GC::new_variable(cs.clone(), || Ok(val.borrow().cmE), mode)?;
+            let cmW = GC::new_variable(cs.clone(), || Ok(val.borrow().cmW), mode)?;
+
+            Ok(Self {
+                _c: PhantomData,
+                cmE,
+                cmW,
+            })
+        })
+    }
+}
+
+/// Implements the circuit that does the checks of the Non-Interactive Folding Scheme Verifier
+/// described in section 4 of https://eprint.iacr.org/2021/370.pdf, where the cmE & cmW checks are
+/// delegated to the NIFSCycleFoldGadget.
+pub struct NIFSGadget<C: CurveGroup> {
+    _c: PhantomData<C>,
+}
+
+impl<C: CurveGroup> NIFSGadget<C>
+where
+    C: CurveGroup,
+{
+    /// Implements the constraints for NIFS.V for u and x, since cm(E) and cm(W) are delegated to
+    /// the CycleFold circuit.
+    pub fn verify(
+        r: FpVar<CF1<C>>,
+        ci1: CommittedInstanceE1Var<C>,
+        ci2: CommittedInstanceE1Var<C>,
+        ci3: CommittedInstanceE1Var<C>,
+    ) -> Result<(), SynthesisError> {
+        ci3.u.enforce_equal(&(ci1.u + r.clone() * ci2.u))?;
+
+        let x_rlc = ci1
+            .x
+            .iter()
+            .zip(ci2.x)
+            .map(|(a, b)| a + &r * &b)
+            .collect::<Vec<FpVar<CF1<C>>>>();
+        x_rlc.enforce_equal(&ci3.x)?;
+
+        Ok(())
+    }
+}
+
+/// NIFSCycleFoldGadget performs the Nova NIFS.V elliptic curve points relation checks in the other
+/// curve following CycleFold (https://eprint.iacr.org/2023/1192.pdf).
+pub struct NIFSCycleFoldGadget<C: CurveGroup, GC: CurveVar<C, CF2<C>>> {
+    _c: PhantomData<C>,
+    _gc: PhantomData<GC>,
+}
+impl<C: CurveGroup, GC: CurveVar<C, CF2<C>>> NIFSCycleFoldGadget<C, GC>
+where
+    C: CurveGroup,
+    GC: CurveVar<C, CF2<C>>,
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    pub fn verify(
+        r_bits: Vec<Boolean<CF2<C>>>,
+        cmT: GC,
+        ci1: CommittedInstanceE2Var<C, GC>,
+        ci2: CommittedInstanceE2Var<C, GC>,
+        ci3: CommittedInstanceE2Var<C, GC>,
+    ) -> Result<(), SynthesisError> {
+        // cm(E) check:
+        ci3.cmE.enforce_equal(
+            &((ci2.cmE.scalar_mul_le(r_bits.iter())? + cmT).scalar_mul_le(r_bits.iter())?
+                + ci1.cmE),
+        )?;
+        // cm(W) check:
+        ECRLC::<C, GC>::check(r_bits, ci1.cmW, ci2.cmW, ci3.cmW)?;
+        Ok(())
+    }
```

Comment:
Same as above. Can we write the exact op we're checking with constraints?

Reply:
Agree, as in https://github.com/privacy-scaling-explorations/folding-schemes/pull/11#discussion_r1315072546 will update it.